gcovr/configuration.py

```diff
@@ -0,0 +1,475 @@
+# Copyright 2013 Sandia Corporation
+# This software is distributed under the BSD license.
+from argparse import ArgumentTypeError, SUPPRESS
+from locale import getpreferredencoding
+from multiprocessing import cpu_count
+import os
+
+
+def check_percentage(value):
+    r""""
+    Check that the percentage is within a reasonable range and if so return it.
+    """
+    try:
+        x = float(value)
+        if not (0.0 <= x <= 100.0):
+            raise ValueError()
+    except ValueError:
+        raise ArgumentTypeError(
+            "{value} not in range [0.0, 100.0]".format(value=value))
+    return x
+
+
+def check_non_empty(value):
+    if not value:
+        raise ArgumentTypeError("value should not be empty")
+    return value
+
+
+class GcovrConfigOption:
+    r"""
+    Represents a single setting for a gcovr runtime parameter.
+
+    Gcovr can be extensively configured through a series of options,
+    representing these options as a simple class object allows them to be
+    portabilty re-used in multiple configuration schemes. This is implemented
+    in a way similar to how options are defined in argparse. The converter
+    keyword argument is expected to return a valid conversion of a string
+    value or throw an error.
+
+    This class is initializer in a very similar fashion as adding arguments
+    to an argparse Parser instance with a few minor changes.
+        1. The first positional argument is the "name" (i.e. "dest" in
+           argparse speak) and is required.
+        2. A "group" keyword argument has been added which corresponds to a
+           key in the GCOVR_CONFIG_OPTION_GROUPS dictionary.
+    """
+
+    def __init__(self, name, *flags, **kwargs):
+        self.name = name
+        self.flags = flags
+        self.group = kwargs.pop("group", None)
+        self.default = None
+        #
+        # set remaining kwargs as attributes
+        for key, value in kwargs.items():
+            setattr(self, key, value)
+        if hasattr(self, 'help'):
+            self.help = self.help.format(**self.__dict__)
+        #
+        # the store_true and store_false actions have hardcoded boolean
+        # constants in their definitions so they need switched to the generic
+        # store_const in order for the logic here to work correctly.
+        if getattr(self, 'action', None) == 'store_true':
+            self.action = 'store_const'
+            self.const = True
+            self.default = False
+        if getattr(self, 'action', None) == 'store_false':
+            self.action = 'store_const'
+            self.const = False
+            self.default = True
+
+    def __repr__(self):
+        r"""String representation of instance."""
+        fmt = "GcovrConfigOption({name!r}, {flags}group={group!r}, {kwargs})"
+        kwargs = dict(self.__dict__)
+        for key in ["name", "flags", "group"]:
+            del kwargs[key]
+        kwargs = ['{k}={v!r}'.format(k=k, v=v) for k, v in kwargs.items()]
```

Comment:
dict iteration order is undefined prior to Python 3.6. To get a stable `repr()` across runs we should sort the entries, e.g. `[... for k, v in sorted(kwargs.items())]`.